### PR TITLE
Feature/spl gen #4

### PR DESCRIPTION
**Pull request by albertmakan**

### CODE REVIEW

File: lang/src/splgen/constants.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+DEPENDENCIES = 'dependencies'
+CODE = 'code'
+BASE = 'base'
+CONTRACTS = 'contracts'
+FUNCTIONS = 'functions'
+MODIFIERS = 'modifiers'
+VARIABLES = 'variables'
+STRUCTS = 'structs'
+EVENTS = 'events'
+GLOBAL = '@global'
```
Comment (panicmilos) [inline lang/src/splgen/constants.py:10]: Smart move 😄

File: lang/src/splgen/generator.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        for ca in admin.contract_administrators:
+            for method in ca.methods:
+                contracts[ca.contract.name].functions[method.name].mod.add(
+                    admin.method.name)
+
+    for contract in contracts.values():
+        for fun in contract.functions.values():
+            fun.insert_modifiers()
+
+    with open(output_file, 'w') as f:
```
Comment (panicmilos) [inline lang/src/splgen/generator.py:64]: Why didn't we generate using jinja?

File: lang/src/splgen/generator.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+
+
+def generate(model: Script, output_file: str):
+    dependencies: Set[Dependency] = set()
+    manager = model.administration.main_administrator
+    contracts = {
+        c.name: Contract(
+            name=c.name,
+            functions={
+                **{
```
Comment (panicmilos) [inline lang/src/splgen/generator.py:19]: What is this magic 😄

File: lang/src/splgen/generator.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+    def __hash__(self):
+        return hash((self.name, self.code, self.type))
+
+
+@dataclass
+class Function:
+    params: Dict[str, str]
+    name: str
+    mod: Set[str]
```
Comment (panicmilos) [inline lang/src/splgen/generator.py:100]: this mod is kinda confusing me tbh.

File: lang/src/splgen/generator.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+                    f"{'Modifier' if fun.is_modifier else 'Function'} '{fun.name}' not found")
+        return self
+
+    def insert_params(self):
+        for fun in self.functions.values():
+            fun.insert_params()
+        return self
+
+    def _resolve(self, fun: Function, globals: Set[Dependency]):
+        pack_name, *path, func_name = fun.name.split('.')
```
Comment (panicmilos) [inline lang/src/splgen/generator.py:148]: You really know python syntax 👏

**Review by panicmilos — APPROVED** [review]
There isn't anything really I can add that will improve this. Looks like you thought of everything, keep the good work going!

File: lang/src/splgen/generator.py (hunk 2 above)
Comment (albertmakan) [reply lang/src/splgen/generator.py:64]: I thought Jinja would be overkill for a single template that has 5 lines.

File: lang/src/splgen/generator.py (hunk 3 above)
Comment (albertmakan) [reply lang/src/splgen/generator.py:19]: Dictionary unpacking

File: lang/src/splgen/generator.py (hunk 4 above)
Comment (albertmakan) [reply lang/src/splgen/generator.py:100]: I will change it to something less confusing

File: lang/src/splgen/generator.py (hunk 5 above)
Comment (albertmakan) [reply lang/src/splgen/generator.py:148]: It's simple

**Pull request merged**